#VRML_SIM R2025a utf8
# CheckerRobot.proto

PROTO CheckerRobot [
  field SFVec3f    translation  0 0 0
  field SFRotation rotation     0 0 1 0
  field SFString   name         "checker_robot"
  field SFColor    color        1 1 1  # Blanco por defecto. (0.1 0.1 0.1 para negro)
  field SFString   controller   "robot_driver"
]
{
  Robot {
    translation IS translation
    rotation IS rotation
    name IS name
    controller IS controller
    children [
      # CUERPO VISUAL
      Solid {
        translation 0 0 0.03
        children [
          Shape {
            appearance PBRAppearance {
              baseColor IS color # Vinculaci√≥n directa, sin errores
              roughness 0.5
              metalness 0.1
            }
            geometry Cylinder {
              height 0.06
              radius 0.08
            }
          }
        ]
        boundingObject Cylinder { height 0.06 radius 0.08 }
        physics Physics { density -1 mass 1.0 }
      }
      
      # CONECTOR PARA REINA
      Connector {
        translation 0 0 0.061
        rotation 0 1 0 1.5708
        name "upper_connector"
        type "magnetic"
        autoLock TRUE
      }
      
      # SENSORES Y MOTORES
      GPS { name "gps" }
      Compass { name "compass" }
      Receiver { name "receiver" channel 1 }
      
      # RUEDA IZQUIERDA
      HingeJoint {
        jointParameters HingeJointParameters { axis 0 1 0 anchor 0 0.06 0.02 }
        device [ RotationalMotor { name "left_motor" } ]
        endPoint Solid {
          translation 0 0.06 0.02
          rotation 1 0 0 1.5708
          children [ DEF W Shape { geometry Cylinder { height 0.01 radius 0.02 } } ]
          boundingObject USE W
          physics Physics {
            density -1 
            mass 1.0 
            damping 
            Damping {
              linear 0.5
              angular 0.5
            }
          }
        }
      }
      
      # RUEDA DERECHA
      HingeJoint {
        jointParameters HingeJointParameters { axis 0 1 0 anchor 0 -0.06 0.02 }
        device [ RotationalMotor { name "right_motor" } ]
        endPoint Solid {
          translation 0 -0.06 0.02
          rotation 1 0 0 1.5708
          children [ USE W ]
          boundingObject USE W
          physics Physics {
            mass 0.05
          }
        }
      }
    ]
    boundingObject Group {
        children [ Transform { translation 0 0 0.03 children [ Cylinder { height 0.06 radius 0.08 } ] } ]
    }
    physics Physics { density -1 mass 1.0 }
  }
}